File: 50-best-practice/Sequences/two_sum.py
def twoSumBruteForce(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: List[int]
    """
    for i, num in enumerate(nums):
        for j, num2 in enumerate(nums):
            if not j == i and num2 == target - num:
                return i, j


def twoSumDict(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: List[int]
    """
    dict_nums = dict()
    for i, num in enumerate(nums):
        remain = target - num
        if remain in dict_nums:
            return dict_nums[remain], i
        dict_nums[num] = i


print(twoSumDict([2, 7, 11, 15], 9))
print(twoSumDict([1, 2, 3, 4, 5, 6, 123, 12, 543, 6546, 11, 22, 33, 44, 55, 66], 70))
print(twoSumDict([3, 3], 6))

File: 50-best-practice-js/Sequences/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSumBruteForce = function (nums, target) {
    for (let i = 0; i < nums.length; i++) {
        for (let j = 1; j < nums.length; j++) {
            if (i !== j) {
                if (nums[j] === target - nums[i]) {
                    return [i,j]
                }
            }
        }
    }
};

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSumHashMap = function (nums, target) {
    let map = new Map()
    for (let i = 0; i < nums.length; i++) {
        let remain = target - nums[i]
        if (map.has(remain)) {
            return [map.get(remain), i]
        }
        map.set(nums[i], i)
    }
};

console.log(twoSumBruteForce([2, 7, 11, 15], 13))
console.log(twoSumHashMap([2, 7, 11, 15], 13))